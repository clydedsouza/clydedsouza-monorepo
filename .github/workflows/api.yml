name: Build and deploy API

on:
  push:
    branches:
      - main
    paths:
      - "api/**"
  pull_request:
    branches:
      - main
    paths:
      - "api/**"
  workflow_dispatch:

jobs:
  build_api:
    name: Build API
    runs-on: ubuntu-latest
    environment:
      name: api
    defaults:
      run:
        shell: bash
        working-directory: api

    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      - name: Install Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18.x

      - name: Install NPM packages
        run: npm ci

      - name: Generate API files
        run: npm run generate:api

      - name: Upload API files
        uses: actions/upload-artifact@v4
        with:
          path: ./api/public
          include-hidden-files: true

  deploy_api:
    name: Deploy API
    runs-on: ubuntu-latest
    environment:
      name: api
    needs: build_api
    if: |
      success() &&
      github.ref == 'refs/heads/main'
    defaults:
      run:
        shell: bash

    steps:
      - name: Download artifact
        uses: actions/download-artifact@v4
        with:
          name: artifact
          path: artifact

      - name: Deploy artifacts to the API repo
        uses: tagus/git-deploy@v0.4.1
        with:
          changes: artifact
          repository: git@github.com:clydedsouza/clydedsouza-web-api.git
          ssh_key: ${{ secrets.API_REPO_DEPLOY_KEY }}
          name: clydedsouza
          email: 16275549+clydedsouza@users.noreply.github.com
          clean_repo: true
          branch: main
          message: "Commit from the monorepo API deploy job"
