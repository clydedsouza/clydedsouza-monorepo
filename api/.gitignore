# Dependency directories
node_modules/

# Optional npm cache directory
.npm

# Exclude temp 'working' directory
public/*.json