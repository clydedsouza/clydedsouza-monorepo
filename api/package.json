{
  "name": "clydedsouza-projects",
  "version": "0.0.1",
  "author": "Clyde D'Souza",
  "scripts": {
    "generate:portfolio": "markdown-json -c ./settings/portfolio.json",
    "generate:projects": "markdown-json -c ./settings/projects.json",
    "generate:books": "markdown-json -c ./settings/books.json",    
    "generate:speaking": "markdown-json -c ./settings/speaking.json",
    "generate:teaching": "markdown-json -c ./settings/teaching.json",
    "generate:highlights": "markdown-json -c ./settings/highlights.json",
    "generate:pinned": "markdown-json -c ./settings/pinned.json",
    "generate:api": "npm run generate:projects & npm run generate:teaching & npm run generate:books & npm run generate:speaking & npm run generate:highlights & npm run generate:pinned & npm run generate:portfolio"
  },
  "dependencies": {
    "markdown-json": "^1.6.1"
  }
}
