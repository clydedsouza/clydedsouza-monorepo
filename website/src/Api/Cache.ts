import { PageTypes } from '../Types/PageTypes'
import { IProjectData } from '../Types/ProjectData'
import { getProjectData } from './Core'

const projectDataCacheMap: Map<PageTypes, Promise<IProjectData>> = new Map()

export const getCachedProjectData = (
  pageType: PageTypes
): Promise<IProjectData> => {
  if (!projectDataCacheMap.has(pageType)) {
    const getProjectDataPromise = getProjectData(pageType).catch((error) => {
      projectDataCacheMap.delete(pageType)
      return Promise.reject(error)
    })
    projectDataCacheMap.set(pageType, getProjectDataPromise)
  }

  return projectDataCacheMap.get(pageType)!
}
