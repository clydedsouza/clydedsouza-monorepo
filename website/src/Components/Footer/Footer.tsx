import Cta from '../Cta/Cta'
import './Footer.scss'

function Footer() {
  return (
    <>
      <footer>
        <Cta />
        <p>Website crafted by Clyde DSouza</p>
      </footer>
    </>
  )
}

export default Footer
