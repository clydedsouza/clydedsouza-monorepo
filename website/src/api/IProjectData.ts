export interface IProjectData {
  data: IProject[]
}

export interface IProject {
  id: string
  title: string
  date: string
  description: string
  website: string
  github: string
  madeUsing: string[]
  category: string
  image: string
  imageDescription: string
}
