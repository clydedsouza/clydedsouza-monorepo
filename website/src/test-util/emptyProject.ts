export const emptyProject = {
    title: '',
    description: '',
    image: '',
    imageDescription: '',
    category: '',
    date: '',
    id: '',
    madeUsing: [],
    website: 'https://test.com',
    github: 'https://github.com',
  }
  