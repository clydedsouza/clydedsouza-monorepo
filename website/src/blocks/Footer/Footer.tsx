import Cta from "@/blocks/Cta/Cta";
import "./Footer.scss";
import { GoogleAds } from "../GoogleAds/GoogleAds";

function Footer() {
  return (
    <>
      <footer>
        <GoogleAds />
        <p>Website crafted by Clyde D&#39;Souza</p>
      </footer>
    </>
  );
}

export default Footer;
