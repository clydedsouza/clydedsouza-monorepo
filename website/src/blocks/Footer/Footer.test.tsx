import * as apiCacheModule from "@/api/Cache";
import { emptyProject } from "@/test-util/emptyProject";
import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import Footer from "./Footer";

jest.mock("../../Api/Cache");

const ctaDataMockResponse = [
  {
    ...emptyProject,
    id: "ai-tell-me-a-story",
    title: "Purchase my book AI, Tell Me a Story from these platforms.",
    website: "https://aitellmeastory.clydedsouza.net",
  },
];

describe("Footer", () => {
  beforeEach(() => {
    jest
      .spyOn(apiCacheModule, "getCachedProjectData")
      .mockImplementationOnce(() =>
        Promise.resolve({
          app: {},
          data: [...ctaDataMockResponse],
        })
      );
  });

  it("should render footer", async () => {
    const { container } = render(<Footer />);
    await waitForElementToBeRemoved(() => screen.queryByRole("progressbar"));
    expect(container).toMatchSnapshot();
  });
});
