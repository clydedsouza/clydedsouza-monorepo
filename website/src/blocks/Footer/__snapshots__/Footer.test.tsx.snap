// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer should render footer 1`] = `
<div>
  <footer>
    <div
      class="google-ads"
    >
      <div
        class="google-ads-container-dev"
      />
      <div
        class="text-faint"
      >
        Google ad
      </div>
    </div>
    <p>
      Website crafted by Clyde D'Souza
    </p>
  </footer>
</div>
`;
